### PR TITLE
[SPARK-9351][SQL] remove literals from grouping expressions in Aggregate

### PR DESCRIPTION
**Pull request by cloud-fan**
literals in grouping expressions have no effect at all, only make our grouping key bigger, so we should remove them in Optimizer.

I also make old and new aggregation code consistent about literals in grouping here. In old aggregation, actually literals in grouping are already removed but new aggregation is not. So I explicitly make it a rule in Optimizer.

### CODE REVIEW

**Comment by rxin** [comment]
Do we have a test case to make sure group by still works when there is no group at all?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -315,6 +315,12 @@ class Analyzer(
           }
         )
 
+      // If we have foldable group expressions, remove them.
+      case a: Aggregate if a.resolved && a.groupingExpressions.exists(_.foldable) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Should this be done in the optimizer?

**Comment by marmbrus** [comment]
Also, anything that actually changes code should have a JIRA.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 2]
```diff
@@ -144,14 +144,14 @@ object PartialAggregation {
...
-          groupingExpressions.filter(!_.isInstanceOf[Literal]).map {
+          groupingExpressions.map {
             case n: NamedExpression => (n, n)
             case other => (other, Alias(other, "PartialGroup")())
           }
 
         // Replace aggregations with a new expression that computes the result from the already
         // computed partial evaluations and grouping values.
-        val rewrittenAggregateExpressions = aggregateExpressions.map(_.transformUp {
+        val rewrittenAggregateExpressions = aggregateExpressions.map(_.transformDown {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: In the new aggregation code path, we use `transformDown` to do the rewriting, as we won't have group expressions inside aggregation expression. We should apply it in this old aggregation code path too.

**Comment by cloud-fan** [comment]
I'm not sure if we could remove literal in group expressions, as it will lead to different result if the `spark.sql.retainGroupColumns` is true. cc @rxin

**Comment by rxin** [comment]
cc @yhuai

**Comment by yhuai** [comment]
@cloud-fan I am probably missing the context at here. What's the problem (and why we need to revert #4169)? Do you mean that the old and new aggregation code path have different output schema when `retainGroupColumns` is true (one has literals and one does not literals?)? Can you give an example and explain what is the difference? It is also good to provide details about the problem (what is the symptom and what is the root cause) in the description, which can help other people understand the problem when they look at commit history.

**Comment by cloud-fan** [comment]
@yhuai sorry that I made a mistake at the beginning, actually `retainGroupColumns` has no problem as we will alias literals in dataframe, and literals in aggregation expressions are already allowed. Thus I think we can safely remove literals in grouping expressions.

And I also make old and new aggregation code consistent about literals in grouping here. In old aggregation, we'll remove literals from grouping but new aggregation is not. So I explicitly make it a rule in Optimizer.

**Comment by cloud-fan** [comment]
retest this please.

**Comment by marmbrus** [comment]
Thanks!  Merging to master.

**Pull request closed without merging**